UI Components: Implementation of Notification Slate and Notification …

Amstutz opened the pull request:
The Implementation of the Interface proposed and accepted in https://github.com/ILIAS-eLearning/ILIAS/pull/2208:

Note: The less still needs some love. However, trunk is moving at tremendous speed. So I will rather finish this up in collaboration with the trunk less division (Yvonne, Christoph) as soon as merged.

This Implementation spawns the following rendered output (see: [Slate Notification example](https://github.com/ILIAS-eLearning/ILIAS/compare/trunk...Amstutz:KS/60/notifications_implementation_ui?expand=1#diff-1122d8fca64630e16a0f5f467840efca)):
![image](https://user-images.githubusercontent.com/1866896/68223035-53837680-ffec-11e9-96f2-b46a881b3b1c.png)

The aggregates opened if the title of generic 2 is clicked, looks as follows:
![image](https://user-images.githubusercontent.com/1866896/68223064-6007cf00-ffec-11e9-9e32-ca01dc2712d2.png)

Amstutz committed:
UI Components: Implementation of Notification Slate and Notification Items

klees commented on src/UI/Component/Item/Factory.php:
```diff
@@ -41,4 +41,57 @@ public function standard($title);
+     *     some repository holding the entry which fired the notification.
+     *     Clicking on the close button removes the Notification permanently.
+     *     Exceptions are Notification Items displaying aggregated information.
+     *     In such a case, Clicking on the title displays the list of the Notifications
```
```suggestion
     *     In such a case, clicking on the title displays the list of the Notifications
```

klees commented on src/UI/Component/Item/Factory.php:
```diff
@@ -41,4 +41,57 @@ public function standard($title);
+     *   interaction:
+     *     1: >
+     *       The main interaction offered by clicking on the Notification Items
+     *       title SHOULD open to some repository holding the entry which fired
```
```suggestion
     *       title SHOULD open some repository holding the entry which fired
```

klees commented on src/UI/Component/Item/Factory.php:
```diff
@@ -41,4 +41,57 @@ public function standard($title);
+     * @param \ILIAS\UI\Component\Symbol\Icon\Icon $icon lead icon
+     * @return \ILIAS\UI\Component\Item\Notification
+     */
+    public function notification($title, \ILIAS\UI\Component\Symbol\Icon\Icon $icon);
```
```suggestion
    public function notification($title, \ILIAS\UI\Component\Symbol\Icon\Icon $lead) : Notification;
```

klees commented on src/UI/Component/Item/Factory.php:
```diff
@@ -41,4 +41,57 @@ public function standard($title);
+     *       The purpose of each interaction MUST be clearly labeled by text.
+     * ---
+     * @param string|\ILIAS\UI\Component\Button\Shy|\ILIAS\UI\Component\Link\Standard $title Title of the item holding the main interaction
+     * @param \ILIAS\UI\Component\Symbol\Icon\Icon $icon lead icon
```
```suggestion
```

klees commented on src/UI/Component/Item/Factory.php:
```diff
@@ -41,4 +41,57 @@ public function standard($title);
+     * ---
+     * @param string|\ILIAS\UI\Component\Button\Shy|\ILIAS\UI\Component\Link\Standard $title Title of the item holding the main interaction
+     * @param \ILIAS\UI\Component\Symbol\Icon\Icon $icon lead icon
+     * @return \ILIAS\UI\Component\Item\Notification
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * the description. Note this should only be used, if the content section
+     * needs to hold legacy content that currently does not have a place in the
+     * UI components.
+     * @param \ILIAS\UI\Component\Legacy\Legacy $component
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * needs to hold legacy content that currently does not have a place in the
+     * UI components.
+     * @param \ILIAS\UI\Component\Legacy\Legacy $component
+     * @return self
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+
+    /**
+     * Get the additional content of the item or null.
+     * @return \ILIAS\UI\Component\Legacy\Legacy|null $component
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * Get the additional content of the item or null.
+     * @return \ILIAS\UI\Component\Legacy\Legacy|null $component
+     */
+    public function getAdditionalContent();
```
```suggestion
    public function getAdditionalContent() : ?\ILIAS\UI\Component\Legacy\Legacy;
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * Get an Item like this with an url to consulted async, when to close button is pressed.
+     * With this url, information may be stored persistently in the DB without interrupting the workflow
+     * of the user (e.g. setting a flag, that the message was consulted).
+     * @param string $url
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * With this url, information may be stored persistently in the DB without interrupting the workflow
+     * of the user (e.g. setting a flag, that the message was consulted).
+     * @param string $url
+     * @return self
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+
+    /**
+     * Get the url attached to this Notification Item
+     * @return self
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * Get the url attached to this Notification Item
+     * @return self
+     */
+    public function getCloseAction();
```
```suggestion
    public function getCloseAction() : Notification;
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * Get an Notification Item like this, but with a set of
+     * Notifications, this Notification Item will aggregate.
+     * @param Notification[] $aggregate_notifications
+     * @return self
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+     * @param \ILIAS\UI\Component\Symbol\Icon\Icon $icon lead icon
+     * @return Icon
+     */
+    public function withLeadIcon(\ILIAS\UI\Component\Symbol\Icon\Icon $icon);
```
```suggestion
    public function withLeadIcon(\ILIAS\UI\Component\Symbol\Icon\Icon $lead) : Notification;
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+
+    /**
+     * Set icon as lead
+     * @param \ILIAS\UI\Component\Symbol\Icon\Icon $icon lead icon
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Set icon as lead
+     * @param \ILIAS\UI\Component\Symbol\Icon\Icon $icon lead icon
+     * @return Icon
```
```suggestion
```

klees commented on src/UI/Component/Item/Notification.php:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Get icon as lead. Note that Notifications only accept Icons as lead,
+     * this is different from the standard Item.
+     * @return \ILIAS\UI\Component\Symbol\Icon\Icon
```
```suggestion
```

klees commented on src/UI/Component/MainControls/Slate/Factory.php:
```diff
@@ -69,4 +69,63 @@ public function combined(
+     *     Notifications Slates are used by the system to publish information to
+     *     the user in the form of Notification Items. The aim of the Notification
+     *     Slates and the Notification Items they contain, is to make notifications
+     *     visible and quickly accessible. They form a centralized channel into
```
```suggestion
     *     visible and quickly accessible. They form a centralized channel which bundles notifications.
```

klees commented on src/UI/Component/MainControls/Slate/Factory.php:
```diff
@@ -69,4 +69,63 @@ public function combined(
+     *     the user in the form of Notification Items. The aim of the Notification
+     *     Slates and the Notification Items they contain, is to make notifications
+     *     visible and quickly accessible. They form a centralized channel into
+     *     which notifications are bundled into. Note that the Notification Slates
```
```suggestion
     *     Note that the Notification Slates
```

klees commented on src/UI/Component/MainControls/Slate/Factory.php:
```diff
@@ -69,4 +69,63 @@ public function combined(
+     *     page loading (like "You have received 1 Contact Request") currently called "toasts".
+     *   composition: >
+     *     Notifications Slates hold Notification Items, displaying information
+     *     on and possible interactions with the displayed Notifications. They
```
```suggestion
     *     on and possible interactions with the displayed notifications. They
```

klees reviewed (changes requested):
Hi @Amstutz!

Thanks for moving the Notifications into the realm of the UI framework. We are moving closer to completion...

I noticed some problems in the KS-entries for the various items that slipped my attention when reviewing the (already accepted) interface the first time. Sry. I would appreciate if we could fix these problems soon. Some of the rules for the usage of the Notification Slate to me clearly demand a concept for notifications beyond mere visuals. I'm curious to see where this will end some day.

Please answer the following questions:
- [x] [Why are Notification Slates engaged by default?](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-db35e3346a8382cbfe421d16f795dfedR91) To me this sentence means that the complete Notification Slate is visible when I open a page, which imo clearly should not be the case. I therefore doubt that this is what you meant to express, so I think there is some wording problem here.
- [x] Do we really want to use vertical alignment of `=` (like [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-14f98e24905e20c4bb653702c662bd8eR104))? To me this looks absolutely horrible in most situations and I doubt that it helps the reader in almost all cases. I mean, look [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-14f98e24905e20c4bb653702c662bd8eR86). It just looks like someone punched a whole in the code and its really hard to visually connect the variable and (              ) the value. Another really odd-looking example is [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-0324fe5f9b0fdd4aaee771ae2e6b1608R4).
- [x] I do not really understand how the JS is supposed to work. It only connects Notification Items and their Close Buttons via positions in the DOM, right? This would be very brittle, imo. I'm not sure if we want to work on this atm, though, because, you know, JS-patterns, client-side, maybe typescript, etc. Maybe you could still comment on the JS a little for future reference.
- [x] Could it be that there are [unrelated changes in the less](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-9ecdf20471b0518df9681dd6a7409c1aR41)?

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] Remove doc-strings in favour of typehints where possible. I suggested a lot of these, but stopped at some point because there were so many and GitHubs JS was slow somehow.
- [x] [rules.usage.1](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-db35e3346a8382cbfe421d16f795dfedR107) is strange. As a service I can only add an `Notification Item`. If I do not have any notifications at a given moment I would need to add some empty notification, right? Per rules of `Notification Item` this Item needs to be removable by clicking close. This is contradictory.

Please implement the following changes:
- [x] Remove this silent but breaking [interface change in `Slate`](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-4fb98be82c321f9af986b5d608ea0b09R22). Slates must [have a symbol](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/UI/Component/MainControls/Factory.php#L213). I think that the `Notification Glyph` is appropriate for the `Notification Slate` and the implementation can just pass that `Symbol` to the constructor of the parent class [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-ec2fe40704f85bbcbcedf239e0455a99R26). Imo it is a bad sign if a child class does not call the parent classes constructor, btw. You will duplicate code, possibly break invariants or miss initialisations now or in the future and possibly also cause other kinds of havoc. This basically is what the "Open/Closed-Principle" from SOLID in my understanding is about.
- [x] Please use a flat if/elseif/../else [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-89ba485a9e7542c93a4db1e8e12f2d9bR28) and [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-89ba485a9e7542c93a4db1e8e12f2d9bR19).
- [x] Please undo the changes in [the `Slate` implementation](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-c0f7db1c2a5b99f27f2fa68e37216c4a). It's (mostly) only whitespace changes (and the breaking interface change). Removing the lines between the various properties for me hurts readability (as well as the vertical `=`-alignment).
- [x] Adjust the names of the methods in the tests. They use snake_case instead of dromedarCase [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-f11bc70c598113bcd3f1113eaf223020).

If you need to discuss anything, feel free to contact me, as always.

Best regards!
@klees

klees reviewed (changes requested):
Hi @Amstutz!

Thanks for moving the Notifications into the realm of the UI framework. We are moving closer to completion...

I noticed some problems in the KS-entries for the various items that slipped my attention when reviewing the (already accepted) interface the first time. Sry. I would appreciate if we could fix these problems soon. Some of the rules for the usage of the Notification Slate to me clearly demand a concept for notifications beyond mere visuals. I'm curious to see where this will end some day.

Please answer the following questions:
- [ ] [Why are Notification Slates engaged by default?](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-db35e3346a8382cbfe421d16f795dfedR91) To me this sentence means that the complete Notification Slate is visible when I open a page, which imo clearly should not be the case. I therefore doubt that this is what you meant to express, so I think there is some wording problem here.
- [ ] Do we really want to use vertical alignment of `=` (like [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-14f98e24905e20c4bb653702c662bd8eR104))? To me this looks absolutely horrible in most situations and I doubt that it helps the reader in almost all cases. I mean, look [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-14f98e24905e20c4bb653702c662bd8eR86). It just looks like someone punched a whole in the code and its really hard to visually connect the variable and (              ) the value. Another really odd-looking example is [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-0324fe5f9b0fdd4aaee771ae2e6b1608R4).
- [ ] I do not really understand how the JS is supposed to work. It only connects Notification Items and their Close Buttons via positions in the DOM, right? This would be very brittle, imo. I'm not sure if we want to work on this atm, though, because, you know, JS-patterns, client-side, maybe typescript, etc. Maybe you could still comment on the JS a little for future reference.
- [ ] Could it be that there are [unrelated changes in the less](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-9ecdf20471b0518df9681dd6a7409c1aR41)?

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [ ] Remove doc-strings in favour of typehints where possible. I suggested a lot of these, but stopped at some point because there were so many and GitHubs JS was slow somehow.
- [ ] [rules.usage.1](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-db35e3346a8382cbfe421d16f795dfedR107) is strange. As a service I can only add an `Notification Item`. If I do not have any notifications at a given moment I would need to add some empty notification, right? Per rules of `Notification Item` this Item needs to be removable by clicking close. This is contradictory.

Please implement the following changes:
- [ ] Remove this silent but breaking [interface change in `Slate`](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-4fb98be82c321f9af986b5d608ea0b09R22). Slates must [have a symbol]/(https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/UI/Component/MainControls/Factory.php#L213). I think that the `Notification Glyph` is appropriate for the `Notification Slate` and the implementation can just pass that `Symbol` to the constructor of the parent class [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-ec2fe40704f85bbcbcedf239e0455a99R26). Imo it is a bad sign if a child class does not call the parent classes constructor, btw. You will duplicate code, possibly break invariants or miss initialisations now or in the future and possibly also cause other kinds of havoc. This basically is what the "Open/Closed-Principle" from SOLID in my understanding is about.
- [ ] Please use a flat if/elseif/../else [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-89ba485a9e7542c93a4db1e8e12f2d9bR28) and [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-89ba485a9e7542c93a4db1e8e12f2d9bR19).
- [ ] Please undo the changes in [the `Slate` implementation](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-c0f7db1c2a5b99f27f2fa68e37216c4a). It's (mostly) only whitespace changes (and the breaking interface change). Removing the lines between the various properties for me hurts readability (as well as the vertical `=`-alignment).
- [ ] Adjust the names of the methods in the tests. They use snake_case instead of dromedarCase [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-f11bc70c598113bcd3f1113eaf223020).

If you need to discuss anything, feel free to contact me, as always.

Best regards!
@klees

Amstutz commented:
Hi @klees 

> I noticed some problems in the KS-entries for the various items that slipped my attention when reviewing the (already accepted) interface the first time. Sry. I would appreciate if we could fix these problems soon. 

A further iteration is certeanly needed as soon as we tackled the notification center as a whole.

> Some of the rules for the usage of the Notification Slate to me clearly demand a concept for notifications beyond mere visuals. I'm curious to see where this will end some day.

I see your point of revising this as soon as we have a general notification concept. However, this here was the best I could come up with at the time. Some issue that seems to have been popping up several times in your feedback is, that you confused the combined slate containing the notification slates with the notification slates the self. I tried to answer your questions bellow.

> Why are Notification Slates engaged by default? To me this sentence means that the complete Notification Slate is visible when I open a page, which imo clearly should not be the case. I therefore doubt that this is what you meant to express, so I think there is some wording problem here.

No this works as intended. They are all visible by default. Note that they are packed into a combined slate, which is not visible by default. That they are, is given by the concept, see: https://docu.ilias.de/goto_docu_wiki_wpage_5118_1357.html

> Do we really want to use vertical alignment of = (like here)? To me this looks absolutely horrible in most situations and I doubt that it helps the reader in almost all cases. I mean, look here. It just looks like someone punched a whole in the code and its really hard to visually connect the variable and ( ) the value. Another really odd-looking example is here.

I do not have a strong opinion. But yes, it seems we do want that:
![image](https://user-images.githubusercontent.com/1866896/68767218-9d89df00-0620-11ea-876b-b8aa440e9e12.png)


> I do not really understand how the JS is supposed to work. It only connects Notification Items and their Close Buttons via positions in the DOM, right? This would be very brittle, imo. I'm not sure if we want to work on this atm, though, because, you know, JS-patterns, client-side, maybe typescript, etc. Maybe you could still comment on the JS a little for future reference.

Yes and no. The ID's used by the notification Items. The use of classes for further identification of the elements inside the component, as long as they are clearly identifiable is IMO fine an is done a lot (e.g. see mainbar or metabar js). Also note that the js is a lot better commented I already did comment the "public" interfac, so what is missing? However Ill try to drop a few more words on the registerAggregatesToggle, if the issues have to do with this function.

>  Could it be that there are unrelated changes in the less?

Define unrelated, IMO no. They should all concern Notifications and their placement. Note that I stumbled over several not very pretty things in there. The worst of which I corrected. The meta bar width was hardcoded. It is needed by the notification slate. One could argue about the counter.less, which mostly corrects a bug IMO which appears in the notifications. maybe you mean that?

>  Remove doc-strings in favour of typehints where possible. I suggested a lot of these, but stopped at some point because there were so many and GitHubs JS was slow somehow.

Fan of typehints, shame on me if they have been missing in many places. Not so much a fan of removing doc-strings. But Ill do, I seem to recall that I have been overruled there somewhen in the past.

>  rules.usage.1 is strange. As a service I can only add an Notification Item. If I do not have any notifications at a given moment I would need to add some empty notification, right? Per rules of Notification Item this Item needs to be removable by clicking close. This is contradictory.

IMO no, the service can send notifications and the user can remove them. I have no strong opinion here. This has been taken from the concept linked above. 

>  Remove this silent but breaking interface change in Slate. Slates must have a symbol. I think that the Notification Glyph is appropriate for the Notification Slate and the implementation can just pass that Symbol to the constructor of the parent class here. 

You seem to confuse the combined slate containing the notification slates with the notification slates itself. Notification slates do not contain a symbol. This should be clear from the interface description: 

> rivals:
> Combined Slates: >
> Combined Slates can hold Bulky Links and other Slates, Notification Slates may only contain Notification Items. Further Combined Slates always require an icon and the contained slates are by default dis-engaged.

I did and do not consider that a change in the interface rather fixing this point contradicting with the presented description.

> Imo it is a bad sign if a child class does not call the parent classes constructor, btw. You will duplicate code, possibly break invariants or miss initialisations now or in the future and possibly also cause other kinds of havoc. This basically is what the "Open/Closed-Principle" from SOLID in my understanding is about.

You are right. Ill move the symbol code out of the constructor of the base abstract class to combined and legecy. It was lazy not to do it in the first place.

>  Please use a flat if/elseif/../else here and here.
Hmm, again code format provided by our default template seems to do that for me, if I use if else... Should I really change?

>  Please undo the changes in the Slate implementation. It's (mostly) only whitespace changes (and the breaking interface change). Removing the lines between the various properties for me hurts readability (as well as the vertical =-alignment).

Ill remove the lines. However, not the extra intends, they seem to have been generated by the official template, which I will continue to use when changing files. 

>  Adjust the names of the methods in the tests. They use snake_case instead of dromedarCase here.

Hmm, Ill do, however, we should really do that in one go one time. I used snake_case there seems all the item test code so fare used the snake case. Seemed of to suddenly change there on the fly. However, definitely see your point, so I change the newly added names here.

Do the provided answers help? To sum up, Ill do the following:

- [x] Try to improve some better JS descriptions.
- [x] Use typehints wherever possible and (sadly) remove the PHP-Docs
- [x] Move the symbol stuff out of the base slate constructor and use this by the notification slate.
- [x] Adjust the methods in the tests from snake_case to camel case.
- [x] Revert changes done by the coding template if you insist on it. Please let me know soon, if you do.
- [x] Remove extra lines in slate.

Fine with that?

Best regards!
@klees

Amstutz commented:
We (@klees and @Amstutz ) just discussed the key points above via skype:
- [x] We agreed the we stick with the initial and decided upon interface, that the slates do contain a Symbol, including the Notification Slate.
- [ ] We agreed that the setting multilign assignment does not comply with our coding style guidelines and will correct that.
- [ ] Concerning the test function naming, we agreed that we need to add this to our todo list in our roadmap (changing snake to camel case).

Amstutz committed:
UI Components: Notification Items, improving JS descriptions

Amstutz committed:
UI Components: Notifications Typo

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>

Amstutz committed:
UI Components: Notifications Typo

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>

Amstutz committed:
UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>

Amstutz committed:
UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de> (+14 squashed commits)
Squashed commits:
[1a40d2d] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[66167b1] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[fcef495] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[37aebe8] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[59ababc] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[a531d14] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[fa51882] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[fab3c0f] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[54b7a29] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[5a86e17] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[e8e1482] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[8c41ace] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[c9906e0] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[2161c076] UI Components: Notifications Remove redundant PHPDocs

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>

Amstutz committed:
UI Components: Notifications word smithing

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de> (+2 squashed commits)
Squashed commits:
[66127b3] UI Components: Notifications word smithing

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>
[d0b4f2e] UI Components: Notifications word smithing

Co-Authored-By: Richard Klees <richard.klees@concepts-and-training.de>

Amstutz committed:
UI Components: Notification Items, fixing type hints and doc strings

Amstutz committed:
UI Components: Reverting Spaces

Amstutz committed:
UI Components: Reverting Spaces II

Amstutz committed:
UI Components: Reverting Spaces III

Amstutz committed:
UI Components: Notifications fix color

Amstutz committed:
UI Components fix cases in test

Amstutz committed:
UI Components: Refactored Symbol usage

Amstutz commented:
Hi @klees 

IMO I implemented the required changes in this PR. As soon as fine, give me a bump, then Ill resolve lasting conflicts and merge - OR - give me the list of the next change requirements ;-)

Note: Thx for the feedback, the slight refactoring concerning the symbol IMO was right, feels cleaner now. Further Unit tests saved my ass today, saving me from committing failing code. Nice to see the use of them while refactoring.

klees commented:
Hi @Amstutz!

Thanks for working on this. I much appreciate the positive feedback about the symbol, thx!

I still can find some of the [odd alignement stuff](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-14f98e24905e20c4bb653702c662bd8eR86).

Please add the snake_case stuff to our Roadmap.

If you could spend some extra minutes it would be nice to our successors to squash the sequence of "UI Components: Notifications Remove redundant PHPDocs" commits into one.

Feel free to merge afterwards.

Thx!

Amstutz commented:
Concerning the alignment, right. But this is my code, and IMO it would be bad to mess with other peoples code, however, I think it is oke and free for everybody to align own code as preferred right. The rule is, do not change - align as you prefer (since we or PSR2 do not make any requirements about it). Did I get this right?

Please add the snake_case stuff to our Roadmap - sure, but this IMO doe not belong here anymore. Note that the snake-case incident is fixed in this PR.

Amstutz committed:
Merge branch 'trunk' into KS/60/notifications_implementation_ui

klees commented:
> But this is my code

Hey, I think this is OUR code =) I mean, we share this among various people, and we could very well agree on (or as coordinators: impose) more than PSR-2 in the context of the UI framework. If we currently don't that's fine, but "this is MY code" imo isn't an argument here.

Amstutz commented:
well yes and no. There is something like personal preference for code and lots of things we agree and commit on following. 

changing existing lines just for a matter of personal preference is bad. To align it to things we committed to is good. You see the diff?

Amstutz committed:
UI Components: Fixing tests due to latest merge.

klees commented:
Go on with this PR, but I promise I will come back with that alignment thing. [These](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-14f98e24905e20c4bb653702c662bd8eR86) [giant](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-89ba485a9e7542c93a4db1e8e12f2d9bR144) [holes](https://github.com/ILIAS-eLearning/ILIAS/pull/2299/files#diff-1122d8fca64630e16a0f5f467840efcaR29) seem to be more than a matter of taste. You might want to have a look at typographic rivers. Maybe I'll convince you some day...
